src/interp.rs

```diff
@@ -19,153 +19,483 @@
-        assert!(self.filter.len() == self.factor);
-        assert!(self.zi < self.delay);
-
-        if src.is_empty() {
-            return;
-        }
-
-        let frames = src.len() / self.channels as usize;
-
-        for (src, (dst, z)) in src.chunks_exact(frames).zip(
-            dst.chunks_exact_mut(frames * self.factor)
-                .zip(self.z.chunks_exact_mut(self.delay)),
-        ) {
-            let mut zi = self.zi;
-
-            for (src, dst) in src.iter().zip(dst.chunks_exact_mut(self.factor)) {
-                // Add sample to delay buffer
-                //
-                // TODO Ringbuffer without bounds checks for z/zi
-                //
-                // Safety: zi is checked to be between 0 and self.delay
-                *unsafe { z.get_unchecked_mut(zi) } = *src;
-
-                // Apply coefficients
-                for (filter, dst) in self.filter.iter().zip(dst.iter_mut()) {
-                    let mut acc = 0.0;
-                    for (c, index) in &*filter.coeff {
-                        let mut i = zi as i32 - *index as i32;
-                        if i < 0 {
-                            i += self.delay as i32;
-                        }
-                        // Safety: zi is checked to be between 0 and self.delay
-                        acc += *unsafe { z.get_unchecked(i as usize) } as f64 * c;
+impl Frame for StereoFrame32 {
+    const CHANNELS: usize = 2;
+
+    #[inline(always)]
+    fn scale_add(&mut self, other: &Self, coeff: f32) {
+        self[0] += other[0] * coeff;
+        self[1] += other[1] * coeff;
+    }
+
+    #[inline(always)]
+    fn from_planar(slice: &[f32], stride: usize) -> Self {
+        [slice[0], slice[stride]]
+    }
+}
+
+impl Frame for SurroundFrame32 {
+    const CHANNELS: usize = 6;
+
+    #[inline(always)]
+    fn scale_add(&mut self, other: &Self, coeff: f32) {
+        self[0] += other[0] * coeff;
+        self[1] += other[1] * coeff;
+        self[2] += other[2] * coeff;
+        self[3] += other[3] * coeff;
+        self[4] += other[4] * coeff;
+        self[5] += other[5] * coeff;
+    }
+
+    #[inline(always)]
+    fn from_planar(slice: &[f32], stride: usize) -> Self {
+        [
+            slice[0],
+            slice[stride],
+            slice[2 * stride],
+            slice[3 * stride],
+            slice[4 * stride],
+            slice[5 * stride],
+        ]
+    }
+}
+
+mod specialized {
+    use super::Frame;
+    use std::f64::consts::PI;
+
+    const ALMOST_ZERO: f64 = 0.000001;
+    const TAPS: usize = 48;
+
+    const FACTOR4: usize = 4;
+    const FACTOR4_INPUT_LENGTH: usize = TAPS / FACTOR4;
+    const FACTOR2: usize = 2;
+    const FACTOR2_INPUT_LENGTH: usize = TAPS / FACTOR2;
+
+    #[derive(Debug)]
+    pub(super) struct Interp4F<F: Frame> {
+        filter: [[f32; FACTOR4]; FACTOR4_INPUT_LENGTH],
+        buffer: [F; FACTOR4_INPUT_LENGTH],
+        buffer_pos: usize,
+    }
+
+    impl<'a, 'b, F> Interp4F<F>
```

Comment:
Can the two variants be generated by a macro or would that be worse readability-wise?

Reply:
They probably can, but TBH I'm not that well-versed with macros. My hope is for const-generics to land soon enough to cover this, and a whole bunch of other repetitive code. (The different Frame declarations comes to mind) If it's all the same to you, perhaps let's just wait for that?

Reply:
Works for me :)